common/recordCreate.js

```diff
@@ -0,0 +1,308 @@
+                    viewModel.submissionButtonDisabled = false;
+
+                    if (exception) AlertsService.addAlert(exception.message, 'error');
+                });
+        }
+
+        /**
+         * addRecords - Function that calls ermrestjs method to add data
+         *
+         * @param  {bool} isUpdate              flag
+         * @param  {object} derivedref          derived referene of the column that is being added
+         * @param  {array} recordEditModel      data array
+         * @param  {bool} isModalUpdate         if updating through record app
+         * @param  {object} rootScope           object contains global values from calling function
+         * @param  {object} onSuccessFunction   callback
+         */
+        function addRecords(isUpdate, derivedref, recordEditModel, isModalUpdate, rootScope, onSuccessFunction) {
+            var model = isModalUpdate ? GV_recordEditModel : recordEditModel;
+            var form = viewModel.formContainer;
+
+            // this will include updated and previous raw values.
+            var submissionRowsCopy = [];
+
+            model.submissionRows.forEach(function(row) {
+                submissionRowsCopy.push(Object.assign({}, row));
+            });
+
+            /**
+             * Add raw values that are not visible to submissionRowsCopy:
+             *
+             * submissionRowsCopy is the datastructure that will be used for creating
+             * the upload url. It must have all the visible and invisible data.
+             * The following makes sure that submissionRowsCopy has all the underlying data
+             */
+            if (isUpdate) {
+                for (var i = 0; i < submissionRowsCopy.length; i++) {
+                    var newData = submissionRowsCopy[i];
+                    var oldData = rootScope.tuples[i].data;
+
+                    // make sure submissionRowsCopy has all the data
+                    for (var key in oldData) {
+                        if (key in newData) continue;
+                        newData[key] = oldData[key];
+                    }
+                }
+            }
+
+            //call uploadFiles which will upload files and callback on success
+            uploadFiles(submissionRowsCopy, rootScope, function() {
+
+                var fn = "create",
+                    args = [submissionRowsCopy];
+                var fnScope = isModalUpdate ? derivedref.unfilteredReference.contextualize.entryCreate : rootScope.reference.unfilteredReference.contextualize.entryCreate;
+
+                if (isUpdate) {
+
+                    var data = checkUpdate(submissionRowsCopy, rootScope);
+
+                    // submit rootScope.tuples because we are changing and
+                    // comparing data from the old data set for the tuple with the updated data set from the UI
+                    fn = "update", fnScope = rootScope.reference, args = [rootScope.tuples];
```

Comment:
Is tuples already an array? I'm curious why it needs to be defined in brackets here?